accepted/00039-content-lifecycle-management.md

```diff
@@ -171,18 +175,113 @@ similar to the schedule definition of a repository sync.
+`sle-15-sp1-pool(packages: kernel, systemd, cron), sle-15-sp1-update(packages: kernel, systemd)`
+
+Channels in the 1st environment (`dev` label):
+`prj1-dev-sle-15-sp1-pool(packages: kernel, systemd), prj1-dev-sle-15-sp1-update(packages: kernel, systemd)` (the `cron` packages was filtered out with a filter)
+
+Create *Snapshot* of the `dev` *Environment*, it contains:
+- references to **source** channels: `sle-15-sp1-pool` and `sle-15-sp1-update`,
+- references to packages contained in the 1st *Environment*: `kernel` and `systemd`.
+
+
+### Restore Content Snapshot
+Restoring a *Snapshot* to a *Environment*.
+
+The data from the *Snapshot* is evaluated and the *Content* is re-created from it (e.g. the software channel tree is re-created in given environment and populated by packages and errata from the *Snapshot*).
+
+Technical note: For this, the adapted code for promoting *Environment*s could be used
+
+### Possible improvement: Using restore functionality for promoting
+The already existing *promote* functionality could be replaced by the *restore* code since it does the same.
+
+### Version management
+Keeping all versions of *Snapshot*s throughout the *Project* existence would be impractical for 2 reasons:
+* DB pollution (growing table for *Snapshot*)
+* integrity issues (it is likely that some entities referenced from very old *Snapshot*s will be deleted).
+
+The proposal is to keep only selected subset of the *Snapshot* versions per each *Environment*:
+* for the "production" *Environment* (= last *Environment*  in the *Path*), keep 3 *Snapshot*s,
+* for other *Environment*s, keep 1 *Snapshot*.
 
-As the current design does not allow a rollback of an *Environment* to previous versions of the *Content Build* the only chance to restore an *Environment* is to use the one from one stage above (restore).
+These numbers can be configurable via a SUSE Manager global option (alternatively, this could be overriden on a *Project* level).
 
-Example scenario:
-* 3 *Environment*s: `DEV`, `TEST`, `PRODUCTION`
-* Promoting *Content Build* from `DEV` to `TEST`
-* Testing the patches on systems in `TEST` fails, now we want to bring back the previous state:
+An alternative proposals:
+- allow users flagging *Environment*s with `important` flag, which would mean 3 snapshots (instead of 1) would be tracked for those, or
+- make the number of kept *Snapshot*s configurable per *Environment*.
 
-The solution can be:
- * 'restore' from `PRODUCTION` to `TEST`
+#### Garbage collection
+Some kind of "garbage collection" has to be run on *Project* operations to remove old *Snapshot* versions. The algorithm makes sure that the existing *Snapshot*s conform to the conditions above (it makes sure that for each *Environment* the wanted number of *Snapshot*s is kept and rest is removed). It should be run on *build*, *promote* and *restore* actions. For this, we need to track the *Snapshot*s used by each *Environment* (M-N relationship).
```

Comment:
@mcalmer I think storing only the latest *Environment* in *Snapshot* as we discussed could be problematic since it lacks the information about order of the snapshots in given environment. We need to be able to reconstruct the order of the snapshots per environment, but the snapshots versions do not need to be strictly in descending order (see the PROD environment in the example below), so we can't rely on those. 
We need such order for 
1. garbage collection, and for
2. displaying the history to the user, so that can revert to "snapshot that was there before").
Keeping track of this with the solution of simple Snapshot -> Environment pointers (N-1) would make this hard.

We could store the information either in an M-N relationship (with explicit order parameter in the linking table) or using some kind of explicit "snapshot path" ("linked list" fashion).

Reply:
See my earlier long comment. At the moment I'm not sure whether you aren't over-engineering things.
But I have a hard time distinguishing between data tables you may need for just keeping history and full versioned environments. The proposal is a bit confusing in that regard.
To repeat my assumption: I don't see an obvious reason why we'd need snapshots per environment in the first place. Snapshots _are_ like environments. At least when you look at the current scope (basically a channel tree with a naming convention plus a database entry that keeps track of the version). As long as you have, let's say, version 3 of your staging, stored _once_, you can use it to get _any_ environment restored to that version.
Let's discuss in person next week. This is hard to get right in GitHub comments. :-)

Reply:
@joesusecom , there are 2 things to differentiate here:
- snapshots
- pointers to snapshots

Environments would only store *pointers* to snapshots, but the actual snapshots and their content won't be duplicated.

Reply:
Ah, that clarifies things. Let's make sure that distinction (storing metadata/pointers vs. duplicating the actual snapshots) becomes clear.